agent/taskresource/envFiles/envfile.go

```diff
@@ -0,0 +1,432 @@
+	"github.com/pkg/errors"
+
+	"github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ioutilwrapper"
+)
+
+const (
+	ResourceName      = "envfile"
+	envFileDirPath    = "envfiles"
+	envTempFilePrefix = "tmp_env"
+
+	s3DownloadTimeout = 30 * time.Second
+)
+
+// EnvironmentFileResource represents envfile as a task resource
+// these environment files are retrieved from s3
+type EnvironmentFileResource struct {
+	cluster     string
+	taskARN     string
+	region      string
+	resourceDir string // path to store env var files
+
+	// env file related attributes
+	environmentFilesSourceUnsafe []apicontainer.EnvironmentFile // list of env file objects
+
+	executionCredentialsID string
+	credentialsManager     credentials.Manager
+	s3ClientCreator        factory.S3ClientCreator
+	os                     oswrapper.OS
+	ioutil                 ioutilwrapper.IOUtil
+
+	// Fields for the common functionality of task resource. Access to these fields are protected by lock.
+	createdAtUnsafe      time.Time
+	desiredStatusUnsafe  resourcestatus.ResourceStatus
+	knownStatusUnsafe    resourcestatus.ResourceStatus
+	appliedStatusUnsafe  resourcestatus.ResourceStatus
+	statusToTransitions  map[resourcestatus.ResourceStatus]func() error
+	terminalReasonUnsafe string
+	terminalReasonOnce   sync.Once
+	lock                 sync.RWMutex
+}
+
+// NewEnvironmentFileResource creates a new EnvironmentFileResource object
+func NewEnvironmentFileResource(cluster, taskARN, region, dataDir string, credentialsManager credentials.Manager,
+	executionCredentialsID string) (*EnvironmentFileResource, error) {
+	envfileResource := &EnvironmentFileResource{
+		cluster:                cluster,
+		taskARN:                taskARN,
+		region:                 region,
+		os:                     oswrapper.NewOS(),
+		ioutil:                 ioutilwrapper.NewIOUtil(),
+		s3ClientCreator:        factory.NewS3ClientCreator(),
+		executionCredentialsID: executionCredentialsID,
+		credentialsManager:     credentialsManager,
+	}
+
+	taskARNFields := strings.Split(taskARN, "/")
+	taskID := taskARNFields[len(taskARNFields)-1]
+	// we save envfiles to path: /var/lib/ecs/data/envfiles/cluster_name/task_id/${s3filename.env}
+	envfileResource.resourceDir = filepath.Join(dataDir, envFileDirPath, cluster, taskID)
+
+	envfileResource.initStatusToTransition()
+	return envfileResource, nil
+}
+
+// Initialize initializes the EnvironmentFileResource
+func (envfile *EnvironmentFileResource) Initialize(resourceFields *taskresource.ResourceFields,
+	taskKnownStatus status.TaskStatus,
+	taskDesiredStatus status.TaskStatus) {
+	envfile.lock.Lock()
+	defer envfile.lock.Unlock()
+
+	envfile.initStatusToTransition()
+	envfile.credentialsManager = resourceFields.CredentialsManager
+	envfile.s3ClientCreator = factory.NewS3ClientCreator()
+	envfile.os = oswrapper.NewOS()
+	envfile.ioutil = ioutilwrapper.NewIOUtil()
```

Comment:
There is one thing here you may want to think about is whether we want to redownload the S3 file when agent crash and restart during resource CREATED but task hasn't go to RUNNING. if we want to redownload, we need to reset the task resource status back to none here.

Reply:
i think you're right. i guess if agent crashes we'd jsut re-download everything again, even if files were previously partially downloaded.